### ISSUE TITLE
Thought: Module Inlining?

### ISSUE THREAD

**Issue by developit** [babel/babili#100]
This preset already seems to do a ton of what other toolchains offer in terms of DCE, mangling, etc.  What about taking that last step and having it support merging a whole graph of ES Modules dependencies into a single closure?

**Comment by kangax** [comment]
Yep, we definitely want to explore some sort of global module/file -aware minification. 

Also related to #62

**Comment by developit** [comment]
@kangax sweet, I was hoping you guys would see the potentially huge value there. The runtime performance benefits of scope flattening are pretty awesome.

**Comment by kangax** [comment]
We're actually planning to do an experiment at Facebook to try flattening scopes and hoist-rename everything to the top level in hopes of gaining on parsing performance. We'll see how it goes.

**Comment by baer** [comment]
@developit - A good friend of mine just announced [Interlock.js](http://interlockjs.com/) in a talk at [Cascadia.js](https://www.youtube.com/watch?v=auKQKaT2qTA) which does just that actually.

It's really interesting actually - it's a pluggable build tool that uses babel for processing the AST just like you're suggesting. Some of the more interesting parts of it are it's bundling system which is based on Git's hashing algorithm to create reproducible module ids which lets you share code chunks across builds. Anyway, just thought I'd share.

**Comment by kangax** [comment]
@baer I really enjoyed that video, thanks! Interesting ideas on bundling and HTTP/2.
